p2p/ipld/plugin/plugin.go

```diff
@@ -0,0 +1,328 @@
+			cid:  block.Cid(),
+			Data: nil,
+		}, nil
+	}
+	domainSeparator := data[:prefixOffset]
+	if bytes.Equal(domainSeparator, leafPrefix) {
+		return &nmtLeafNode{
+			cid:  block.Cid(),
+			Data: data[prefixOffset:],
+		}, nil
+	} else if bytes.Equal(domainSeparator, innerPrefix) {
+		return nmtNode{
+			cid: block.Cid(),
+			l:   data[prefixOffset : prefixOffset+nmtHashSize],
+			r:   data[prefixOffset+nmtHashSize:],
+		}, nil
+	}
+	return nil, fmt.Errorf(
+		"expected first byte of block to be either the leaf or inner node prefix: (%x, %x), got: %x)",
+		leafPrefix,
+		innerPrefix,
+		domainSeparator,
+	)
+}
+
+var _ node.Node = (*nmtNode)(nil)
+var _ node.Node = (*nmtLeafNode)(nil)
+
+type nmtNode struct {
+	cid  cid.Cid
+	l, r []byte
+}
+
+func (n nmtNode) RawData() []byte {
+	return append([]byte{nmt.NodePrefix}, append(n.l, n.r...)...)
+}
+
+func (n nmtNode) Cid() cid.Cid {
+	return n.cid
+}
+
+func (n nmtNode) String() string {
+	return fmt.Sprintf(`
+node {
+	hash: %x,
+	l: %x,
+	r: %x"
+}`, n.cid.Hash(), n.l, n.r)
+}
+
+func (n nmtNode) Loggable() map[string]interface{} {
+	return nil
+}
+
+func (n nmtNode) Resolve(path []string) (interface{}, []string, error) {
+	switch path[0] {
+	case "0":
+		return &node.Link{Cid: cidFromNamespacedSha256(n.l)}, path[1:], nil
+	case "1":
+		return &node.Link{Cid: cidFromNamespacedSha256(n.r)}, path[1:], nil
+	default:
+		return nil, nil, errors.New("invalid path for inner node")
+	}
+}
+
+func (n nmtNode) Tree(path string, depth int) []string {
+	if path != "" || depth == 0 {
+		return nil
+	}
+
+	return []string{
+		"0",
+		"1",
+	}
+}
+
+func (n nmtNode) ResolveLink(path []string) (*node.Link, []string, error) {
+	obj, rest, err := n.Resolve(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	lnk, ok := obj.(*node.Link)
+	if !ok {
+		return nil, nil, fmt.Errorf("was not a link")
+	}
+
+	return lnk, rest, nil
+}
+
+func (n nmtNode) Copy() node.Node {
+	panic("implement me")
+}
+
+func (n nmtNode) Links() []*node.Link {
+	return []*node.Link{{Cid: cidFromNamespacedSha256(n.l)}, {Cid: cidFromNamespacedSha256(n.r)}}
+}
+
+func (n nmtNode) Stat() (*node.NodeStat, error) {
+	return &node.NodeStat{}, nil
+}
+
+func (n nmtNode) Size() (uint64, error) {
+	return 0, nil
+}
+
+type nmtLeafNode struct {
+	cid  cid.Cid
+	Data []byte
+}
+
+func (l nmtLeafNode) RawData() []byte {
+	return append([]byte{nmt.LeafPrefix}, l.Data...)
+}
+
+func (l nmtLeafNode) Cid() cid.Cid {
+	return l.cid
+}
+
+func (l nmtLeafNode) String() string {
+	return fmt.Sprintf(`
+leaf {
+	hash: 		%x,
+	len(Data): 	%v
+}`, l.cid.Hash(), len(l.Data))
+}
+
+func (l nmtLeafNode) Loggable() map[string]interface{} {
+	return nil
+}
+
+func (l nmtLeafNode) Resolve(path []string) (interface{}, []string, error) {
+	if path[0] == "Data" {
+		// TODO: store the data separately
+		// currently nmtLeafNode{Data:} contains the actual data
+		// instead, there should be a link in the leaf to the actual data
+		return nil, nil, nil
```

Comment:
This is the only thing that needs further thinking but could be done in a followup PR.